2015/intro_concurrency.slide

```diff
@@ -0,0 +1,375 @@
+
+* Putting them all together: Monte Carlo
+.play intro_concurrency/mc.go /func main/,$
+
+* Interfaces
+An *interface* is a set of methods
+
+  type Distribution interface {
+      Rand() float64
+  }
+
+A type *satisfies* the inteface if it has those methods
+
+  type Uniform struct{
+      Min float64
+      Max float64
+  }
+
+  func (u Uniform) Rand() float64 {
+      return rand.Float64()*(u.Max - u.Min) + u.Min
+  }
+
+Uniform satisfies the interface _implicitly_
+
+* Monte Carlo with interfaces
+.play intro_concurrency/interface.go /MonteCarlo estimates/,$
+
+* Concurrency
+
+* Concurrency
+
+MPI:
+
+- One source code, but multiple instances running
+- One thread per instance
+- Distributed memory (even in a shared-memory environment)
+- Programmer explicitly describes what happens when
+
+* Concurrency
+Go:
+
+- One source code
+- One program instance
+- Shared memory
+- Programmer describes what may happen at the same time
+- Go runtime decides how it happens
+- [[https://www.youtube.com/watch?v=cN_DpYBzKso][Concurrency is not parallelism]]
+- Does not prohibit MPI for distributed memory computation
+
+* Concurrency
+Three major parts 
+
+- Goroutines (`go` keyword) describe independently executing units
+- Channels (`chan` type) communicate between goroutines
+- Select statement chooses between channels
+
+`sync` package provides synchronization primitives (`sync.Mutex`, `sync.WaitGroup`)
+
+* Goroutines
+The `go` command launches a concurrently executing process
+
+  go func()
+  go func(x, y, ...)
+
+Like a thread, but very cheap
+
+- Goroutines execute independently from one another
+- Millions of goroutines common
+
+Automatically scheduled onto OS threads by the Go runtime
+
+Enable parallelism with
+
+  runtime.GOMAXPROCS(n)
+  runtime.GOMAXPROCS(runtime.NumCPU())  // Will be automatic in Go 1.5
+
+* Multiple goroutine example
+.play intro_concurrency/goroutines_exit.go /func main/,$
+
+Program ends when main returns
+
+* Example -- WaitGroup
+.play intro_concurrency/goroutines.go /func main/,$
+
+* Parallel Monte Carlo
+.play intro_concurrency/parmc.go /func Par/,$
+
+* Channels
+Channels communicate between two goroutines
+
+Write to a channel
+
+   c <- v // Goroutine blocks until can write
+
+Read from a channel
+
+   v := <-c // Goroutine blocks until can read
+
+Unbuffered (like a synchronous send/receive)
+
+  make(chan float64)
+
+Buffered (like an asynchronous send/receive)
+
+  make(chan float64, 10)
+
+* Channel communication
+.play intro_concurrency/chan.go /func main/,$
+
+Notice that the order is the same
+
+* Channels
+Can have multiple producers and consumers
+
+.play intro_concurrency/multichan.go /func main/,$
+
+* Range
+The `range` keyword reads from the channel until it is closed
+
+.play intro_concurrency/chan_range.go /func consumer/,$
+
+* Deadlock Detection (not perfect)
+
+.play intro_concurrency/deadlock.go
+
+Race detector available as well
+  
+  go install -race myprogram.go
```

Comment:
I thought I commented this here, but obviously not.

It's my understanding that you can do a race detector run in a slide by holding shift when you run the code, but I can't get it to work.